pkg/controller/common/defaults/pod_template.go

```diff
@@ -319,29 +249,17 @@ func (b *PodTemplateBuilder) WithInitContainers(initContainers ...corev1.Contain
 // If a zero-value (empty map) for at least one of limits or request is provided, the given resource requirements
 // are not applied: the user may want to use a LimitRange.
 func (b *PodTemplateBuilder) WithResources(resources corev1.ResourceRequirements) *PodTemplateBuilder {
-	if b.Container.Resources.Requests == nil && b.Container.Resources.Limits == nil {
-		b.Container.Resources = resources
-	}
+	b.containerDefaulter.WithResources(resources)
 	return b
 }
 
 func (b *PodTemplateBuilder) WithPreStopHook(handler corev1.Handler) *PodTemplateBuilder {
-	if b.Container.Lifecycle == nil {
-		b.Container.Lifecycle = &corev1.Lifecycle{}
-	}
-
-	if b.Container.Lifecycle.PreStop == nil {
-		// no user-provided hook, we can use our own
-		b.Container.Lifecycle.PreStop = &handler
-	}
-
+	b.containerDefaulter.WithPreStopHook(&handler)
 	return b
 }
 
 func (b *PodTemplateBuilder) WithArgs(args ...string) *PodTemplateBuilder {
-	if b.Container.Args == nil {
-		b.Container.Args = args
-	}
+	b.containerDefaulter.WithArgs(args)
 	return b
 }
 
```

Comment:
Should the `PodTemplateBuilder` just embed a Defaulter to inherit all its functions, so we avoid the extra hop for most functions (one function in PodTemplateBuilder that just calls the corresponding function in Defaulter)?

```
type PodTemplateBuilder struct {
    container.Defaulter
    PodTemplate        corev1.PodTemplateSpec
    containerName      string
}
```

Reply:
If `container.Defaulter` is embedded then the return type of the "containers related" functions becomes `Defaulter`. While all the builders are expecting a `PodTemplateBuilder` to chain the calls.

For instance it breaks [the following code](https://github.com/elastic/cloud-on-k8s/blob/master/pkg/controller/apmserver/pod.go#L106): 

```golang
	builder := defaults.NewPodTemplateBuilder(
		p.PodTemplate, apmv1.ApmServerContainerName).
		WithResources(DefaultResources). # error here 
		WithDockerImage(p.CustomImageName, container.ImageRepository(container.APMServerImage, p.Version)).
		WithReadinessProbe(readinessProbe(as.Spec.HTTP.TLS.Enabled())).
		WithPorts(ports).
		WithCommand(command).
		WithVolumes(configVolume.Volume(), configSecretVolume.Volume()).
		WithVolumeMounts(configVolume.VolumeMount(), configSecretVolume.VolumeMount()).
		WithEnv(env...)


	if p.keystoreResources != nil {
		dataVolume := keystore.DataVolume(
			strings.ToLower(as.Kind),
			DataVolumePath,
		)
		builder.WithInitContainers(p.keystoreResources.InitContainer).
			WithVolumes(p.keystoreResources.Volume, dataVolume.Volume()).
			WithVolumeMounts(dataVolume.VolumeMount()).
			WithInitContainerDefaults()
	}

	return builder.PodTemplate
```

IMHO a `podTemplateBuilder` should not expose those methods at all, but delegate the container setup to a dedicated builder/defaulter.
E.g. maybe something along those lines :

```golang
	builder := defaults.NewPodTemplateBuilder(p.PodTemplate, apmv1.ApmServerContainerName).
		WithVolumes(...). // Pod specific methods
                WithTerminationGracePeriod(...).
                WithServiceAccount(...)
                WithContainer(
                          NewContainerDefaulter("apm-server").
		          WithResources(DefaultResources). // Container specific methods
		          WithDockerImage(.....).
		          WithReadinessProbe(readinessProbe(as.Spec.HTTP.TLS.Enabled())).
		          WithPorts(ports).
		          WithCommand(command).
		          WithVolumeMounts(configVolume.VolumeMount(), configSecretVolume.VolumeMount()).
		          WithEnv(env...).
                          Container()
                )

```

I can create a separate issue to track this refactoring effort.

Reply:
I understand, that makes sense 👍